cmd/xl-zones.go

```diff
@@ -0,0 +1,1068 @@
+		}
+	}
+
+	var isTruncated = false
+	for _, entriesValid := range zoneEntriesValid {
+		for _, valid := range entriesValid {
+			if !valid {
+				continue
+			}
+			isTruncated = true
+			break
+		}
+		if isTruncated {
+			break
+		}
+	}
+
+	var lentry FileInfo
+	var found bool
+	var zoneIndex = -1
+	for i, entriesValid := range zoneEntriesValid {
+		for j, valid := range entriesValid {
+			if !valid {
+				continue
+			}
+			if !found {
+				lentry = zoneEntries[i][j]
+				found = true
+				zoneIndex = i
+				continue
+			}
+			if zoneEntries[i][j].Name < lentry.Name {
+				lentry = zoneEntries[i][j]
+				zoneIndex = i
+			}
+		}
+	}
+
+	// We haven't been able to find any least entry,
+	// this would mean that we don't have valid entry.
+	if !found {
+		return lentry, 0, zoneIndex, isTruncated
+	}
+
+	leastEntryCount := 0
+	for i, entriesValid := range zoneEntriesValid {
+		for j, valid := range entriesValid {
+			if !valid {
+				continue
+			}
+
+			// Entries are duplicated across disks,
+			// we should simply skip such entries.
+			if lentry.Name == zoneEntries[i][j].Name && lentry.ModTime.Equal(zoneEntries[i][j].ModTime) {
+				leastEntryCount++
+				continue
+			}
+
+			// Push all entries which are lexically higher
+			// and will be returned later in Pop()
+			zoneEntryChs[i][j].Push(zoneEntries[i][j])
+		}
+	}
+
+	return lentry, leastEntryCount, zoneIndex, isTruncated
+}
+
+// mergeZonesEntriesCh - merges FileInfo channel to entries upto maxKeys.
+func mergeZonesEntriesCh(zonesEntryChs [][]FileInfoCh, maxKeys int, zoneDrives []int, heal bool) (entries FilesInfo) {
+	var i = 0
+	var zonesEntriesInfos [][]FileInfo
+	var zonesEntriesValid [][]bool
+	for _, entryChs := range zonesEntryChs {
+		zonesEntriesInfos = append(zonesEntriesInfos, make([]FileInfo, len(entryChs)))
+		zonesEntriesValid = append(zonesEntriesValid, make([]bool, len(entryChs)))
+	}
+	for {
+		fi, quorumCount, zoneIndex, valid := leastEntryZone(zonesEntryChs, zonesEntriesInfos, zonesEntriesValid)
+		if !valid {
+			// We have reached EOF across all entryChs, break the loop.
+			break
+		}
+		rquorum := fi.Quorum
+		// Quorum is zero for all directories.
+		if rquorum == 0 {
+			// Choose N/2 quoroum for directory entries.
+			rquorum = zoneDrives[zoneIndex] / 2
+		}
+
+		if heal {
+			// When healing is enabled, we should
+			// list only objects which need healing.
+			if quorumCount == zoneDrives[zoneIndex] {
+				// Skip good entries.
+				continue
+			}
+		} else {
+			// Regular listing, we skip entries not in quorum.
+			if quorumCount < rquorum {
+				// Skip entries which do not have quorum.
+				continue
+			}
+		}
+		entries.Files = append(entries.Files, fi)
+		i++
+		if i == maxKeys {
+			entries.IsTruncated = isTruncatedZones(zonesEntryChs, zonesEntriesInfos, zonesEntriesValid)
+			break
+		}
+	}
+	return entries
+}
+
+func isTruncatedZones(zoneEntryChs [][]FileInfoCh, zoneEntries [][]FileInfo, zoneEntriesValid [][]bool) bool {
```

Comment:
Before this is finished, please add some explanations to functions methods in here. There is a lot going on.

Reply:
Definitely, I wanted to get it in some shape first quickly. Will add comments.